### ISSUE TITLE
MPI tests may fail due to lack of CPUs / missing --oversubscribe

### ISSUE THREAD

**Issue by heplesser** [nest/nest-simulator#1725]
**Describe the bug**
MPI tests in the testsuite may fail if they need more parallel processes than the MPI installation is configured for. This happens easily on laptop computers with 2-4 cores, since many MPI tests need 4, some even more processes.

**To Reproduce**
Execute `make installcheck` on a computer with at most 4 cores and an MPI configuration that does not allow oversubscription. This is, e.g., the case for OpenMPI installed from Homebrew with the default `.nestrc` file.

**Expected behavior**
The testsuite passes, including all MPI tests.

**Screenshots**
See e.g. this example: https://github.com/nest/nest-simulator/pull/1712#issuecomment-676476384

Error output can look like this:
```
  Running test 'mpitests/test_localonly.sli'... Failed: missed SLI assertion
==================================================
Following is the full output of the test:
==================================================
Running '.../NEST/code/bld_clang_mpi/install/bin/sli' '.../NEST/code/bld_clang_mpi/install/share/doc/nest/mpitests/test_localonly.sli'
SLI master@4f2585467 (C) 2004 The NEST Initiative
N_MPI: 1
N_MPI: 2
N_MPI: 4
N_MPI: 8
--------------------------------------------------------------------------
There are not enough slots available in the system to satisfy the 8
slots that were requested by the application:

  .../NEST/code/bld_clang_mpi/install/bin/nest

Either request fewer slots for your application, or make more slots
available for use.

A "slot" is the Open MPI term for an allocatable unit where we can
launch a process.  The number of slots available are defined by the
environment in which Open MPI processes are run:

  1. Hostfile, via "slots=N" clauses (N defaults to number of
     processor cores if not provided)
  2. The --host command line parameter, via a ":N" suffix on the
     hostname (N defaults to 1 if not provided)
  3. Resource manager (e.g., SLURM, PBS/Torque, LSF, etc.)
  4. If none of a hostfile, the --host command line parameter, or an
     RM is present, Open MPI defaults to the number of processor cores

In all the above cases, if you want Open MPI to default to the number
of hardware threads instead of the number of processor cores, use the
--use-hwthread-cpus option.

Alternatively, you can use the --oversubscribe option to ignore the
number of available slots when deciding the number of processes to
launch.
--------------------------------------------------------------------------
```

**Desktop/Environment (please complete the following information):**
 - OS: macOS 10.15.6
 - Python-Version: Python 3.8
 - NEST-Version:  master@4f2585467
 - Installation: 
```
--------------------------------------------------------------------------------
NEST Configuration Summary
--------------------------------------------------------------------------------

Build type          : 
Target System       : Darwin
Cross Compiling     : FALSE
C compiler          : AppleClang 11.0.3.11030032 (/Library/Developer/CommandLineTools/usr/bin/cc)
C compiler flags    :  -O2 -Wall -Xclang -fopenmp    
C++ compiler        : AppleClang 11.0.3.11030032 (/Library/Developer/CommandLineTools/usr/bin/c++)
C++ compiler flags  :  -std=c++11 -O2 -Wall -Xclang -fopenmp  
Build dynamic       : ON

Built-in modules    : models;topology
User modules        : None
Python bindings     : Yes (Python 3.8.5: .../opt/miniconda3/envs/nest_dev/bin/python3)
       Includes     : .../opt/miniconda3/envs/nest_dev/include/python3.8
       Libraries    : .../opt/miniconda3/envs/nest_dev/lib/libpython3.8.dylib

Cython bindings     : Yes (Cython 0.29.21: .../opt/miniconda3/envs/nest_dev/bin/cython)
MPI4PY header	    : NO 
Use threading       : Yes (OpenMP: -Xclang -fopenmp)
Use GSL             : Yes (GSL 2.6)
    Includes        : /usr/local/Cellar/gsl/2.6/include
    Libraries       : /usr/local/Cellar/gsl/2.6/lib/libgsl.dylib;/usr/local/Cellar/gsl/2.6/lib/libgslcblas.dylib

Use Readline        : Yes (GNU Readline 4.2)
    Includes        : /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include
    Libraries       : /usr/lib/libreadline.dylib;/usr/lib/libncurses.dylib

Use libltdl         : Yes (LTDL 2.4.6)
    Includes        : /usr/local/include
    Libraries       : /usr/local/lib/libltdl.dylib

Use doxygen         : No
Use MPI             : Yes (MPI: /usr/local/bin/mpicxx)
    FLAGS           : 
    Includes        : /usr/local/Cellar/open-mpi/4.0.4_1/include
    Link Flags      : -L/usr/local/Cellar/libevent/2.1.12/lib
    Libraries       : /usr/local/Cellar/open-mpi/4.0.4_1/lib/libmpi.dylib

Use MUSIC           : No
Use libneurosim     : No
Use Boost           : Yes (Boost 1.73.0)
    Includes        : /usr/local/include
    Libraries       : 

Use recording backend Arbor   : No
Use SIONlib         : No
```

**Additional context**
The work-around for this problem is to add the `--oversubscribe` option to the `.nestrc` file so it looks like this:
```
/mpirun
[/integertype /stringtype /stringtype]
[/numproc     /executable /scriptfile]
{
 () [
  (mpirun --oversubscribe -np ) numproc cvs ( ) executable ( ) scriptfile
 ] {join} Fold
} Function def
```

The key questions are as follows:
- Should we add `--oversubscribe` to `mpirun` by default?
- Should we add an option to the SLI `mpirun` command that activates `--oversubscribe`, e.g. when using `mpirun` from the testssuite?
- Will `--oversubscribe` work for other MPI libraries than `OpenMPI`?

**Comment by jougs** [comment]
> * Should we add --oversubscribe to mpirun by default?

I don't think so. Old versions of at least OpenMPI don't support this

> * Should we add an option to the SLI mpirun command that activates --oversubscribe, e.g. when using mpirun from the testssuite?

See above. On Travis, we are already adding `--oversubscribe` for the builds that need it.

> * Will --oversubscribe work for other MPI libraries than OpenMPI?

I don't know, but don't think so.

Aren't cases like these exactly the reason we made the SLI `mpirun` command configurable by the user? We won't ever be able to set it correctly for *all* users. In this specific instance of the problem, the implementation even tells the user exactly what the problem is and how to circumvent it.

So, in summary, I would tend to consider this a user error and close the issue *wontfix*.
